### ISSUE TITLE
speed up builds

### ISSUE THREAD

**Issue by daler** [bioconda/bioconda-recipes#597]
This is about the structure of `build-packages.py` and how we might speed up builds on Travis-CI.

Currently `build-packages.py` only removes a recipe from the list to build if *all* builds (py27, py34, py35) exist on conda. This means that even if a recipe is known to fail on py27 and therefore `skip: True #[py27]` is already specified, that recipe will still get added to the list. The result is a lot of metadata inspection-and-then-skipping, which in the travis log looks like this:

```
Fetching package metadata: ............
bx-python-0.7.3-py27_0 is already built, skipping.
Fetching package metadata: ............
Skipped: The bx-python-0.7.3-py34_0 recipe defines build/skip for this configuration.
Fetching package metadata: ............
Skipped: The bx-python-0.7.3-py35_0 recipe defines build/skip for this configuration.
...
(etc)
```

For the current set of skipped recipes, this takes ~4mins for what is essentially a no-op. That's the lower bound of how long a single test will take. Waiting for builds was brought up in #595; I think this is the primary issue.

I propose a more careful check of the `Metadata` object in `build-packages.py` that explicitly loops through the Python versions and also keeps track of the current OS so that *only* the recipes that need to get built make it into the queue. I think the best approach will be to restructure the build scripts to use travis-ci's build matrix strategy for the combinations of (linux, osx) and (py27, py34, py35) since on the travis-ci free tier this is pretty much the only way to parallelize builds.

While we're at it, it would be helpful to have helper functions (getting, filtering, toposorting recipes) available as a module that the ucsc and bioconductor recipe scripts could use. #596 is a start of this in case it's useful.

**Comment by johanneskoester** [comment]
Sounds good!

**Comment by daler** [comment]
I think everything here has been resolved in #1436, #1371 (thanks @johanneskoester for those elegant improvements!), and the last part in #1439
